### PR TITLE
Split Iter & IterByColRange types into Tx and MutTxId versions

### LINKED ISSUE #2024: perf: split `Iter` & `IterByColRange` types into `Tx` and `MutTxId` versions
We're wasting work for the `Tx` version doing unnecessary bookkeeping in case there is a `TxState`, but there will never be one.
Meanwhile, the `MutTxId` version is also doing some minor bookkeeping for the case there won't be a `TxState`, but there always will be one.

In the case of `Iter` for `TxId`, this can just be `TableScanIter`, which is about 4x faster according to flamegraphs.

### PR DESCRIPTION
**Pull request by mamcx**
# Description of Changes

Refactor `Iter / CommittedIndexIter` for `Tx` & `MutTx` variants, and solves early the logic for decide the state machine.

Closes #2024

# Expected complexity level and risk

1

# Testing

*Describe any testing you've done, and any testing you'd like your reviewers to do,
so that you're confident that all the changes work as expected!*

- [x] *Run benches with `sudo nice -n -20 taskpolicy -B -t 5 -l 5 -c utility cargo bench --bench subscription --  --save-baseline subs`*

### CODE REVIEW

**Review by Centril — COMMENTED** [review]
Almost there, this is looking very promising in terms of performance improvements.

File: crates/core/src/db/datastore/locking_tx_datastore/state_view.rs [hunk 1]
```diff
@@ -330,48 +358,103 @@ impl<'a> Iterator for IndexSeekIterMutTxId<'a> {
...
+    range: R,
+}
+
+impl<'a, R: RangeBounds<AlgebraicValue>> ScanIterTxByColRange<'a, R> {
+    pub(super) fn new(scan_iter: IterTx<'a>, cols: ColList, range: R) -> Self {
+        Self { scan_iter, cols, range }
+    }
+}
+
+impl<'a, R: RangeBounds<AlgebraicValue>> Iterator for ScanIterTxByColRange<'a, R> {
```
[the remark this answers is not in the document]
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/state_view.rs:428]: Better to do when the new query lands...

File: crates/core/src/db/datastore/locking_tx_datastore/committed_state.rs [hunk 2]
```diff
@@ -628,42 +646,61 @@ impl CommittedState {
...
     committed_rows: IndexScanIter<'a>,
     num_committed_rows_fetched: u64,
 }
 
-impl<'a> CommittedIndexIter<'a> {
-    pub(super) fn new(
-        table_id: TableId,
-        tx_state: Option<&'a TxState>,
-        committed_state: &'a CommittedState,
-        committed_rows: IndexScanIter<'a>,
```
[the remark this answers is not in the document]
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/committed_state.rs:645]: Done. I also split `IndexSeekIterMutTxId`

File: crates/core/src/db/datastore/locking_tx_datastore/state_view.rs [hunk 3]
```diff
@@ -318,60 +358,124 @@ impl<'a> Iterator for IndexSeekIterMutTxId<'a> {
...
             // As a result, in MVCC, this branch will need to check if the `row_ref`
             // also exists in the `tx_state.insert_tables` and ensure it is yielded only once.
-            .and_then(|i| i.find(|row_ref| !self.tx_state.is_deleted(self.table_id, row_ref.pointer())))
+            .and_then(|i| i.find(|row_ref| !self.del_table.contains(&row_ref.pointer())))
         {
             // TODO(metrics): This doesn't actually fetch a row.
             // Move this counter to `RowRef::read_row`.
-            self.num_committed_rows_fetched += 1;
+            // self.num_committed_rows_fetched += 1;
             return Some(row_ref);
```
Comment (mamcx) [inline crates/core/src/db/datastore/locking_tx_datastore/state_view.rs:366]: I intentionally keep this as a reminder, I have no context to say why it is there or how to change it.

File: crates/core/src/db/datastore/locking_tx_datastore/committed_state.rs [hunk 4]
```diff
@@ -628,42 +646,61 @@ impl CommittedState {
...
+}
+
+pub struct CommittedIndexIterMutTx<'a> {
     table_id: TableId,
-    tx_state: Option<&'a TxState>,
-    #[allow(dead_code)]
-    committed_state: &'a CommittedState,
+    tx_state: &'a TxState,
     committed_rows: IndexScanIter<'a>,
     num_committed_rows_fetched: u64,
```
[the remark this answers is not in the document]
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/committed_state.rs]: Done

File: crates/core/src/db/datastore/locking_tx_datastore/datastore.rs [hunk 5]
```diff
@@ -323,15 +324,26 @@ impl Tx for Locking {
...
+    type IterMutByColRange<'a, R: RangeBounds<AlgebraicValue>>
+    = IterMutTxByColRange<'a, R>
+    where
+        Self: 'a;
+    type IterByColEq<'a, 'r>
+        = IterTxByColRange<'a, &'r AlgebraicValue>
+    where
+        Self: 'a;
+    type IterMutByColEq<'a, 'r>
+    = IterMutTxByColRange<'a, &'r AlgebraicValue>
```
[the remark this answers is not in the document]
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/datastore.rs]: Done

[the remark this answers is not in the document]
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/datastore.rs]: It is define as `MutTxDatastore: TxDatastore + MutTx `, so it get the names from `TxDatastore`

[the remark this answers is not in the document]
Comment (mamcx) [reply crates/core/src/db/datastore/locking_tx_datastore/datastore.rs]: Ok, thanks for the clarification. Is done.

**Review by Centril — APPROVED** [review]
Looks good now :)

**Review by cloutiertyler — APPROVED** [review]
I've reviewed just what I am the code owner of. The changes to the datastore trait look good to me.

**Pull request merged**